<a href="https://colab.research.google.com/github/jmmarinr/ComputationalMethods/blob/master/Algebra_Lineal/Complejas_ValoresPropios.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#@title Librerias

import numpy as np
import numpy.linalg as la
import scipy.linalg as sla
import matplotlib.pyplot as plt
import sympy as sp

---
### Profesor: Juan Marcos Marín
---

## Norma de una matriz

En álgebra lineal y análisis numérico, la **norma de una matriz** es una medida que cuantifica el "tamaño" o "magnitud" de una matriz. Existen diferentes normas, cada una con aplicaciones específicas. Las más comunes son:

### 1. **Norma Inducida por un Vector**
Sea $A \in \mathbb{R}^{m \times n}$, y $\|\cdot\|$ una norma vectorial. La norma inducida de $A$ está definida como:

$$
\|A\| = \sup_{x \neq 0} \frac{\|Ax\|}{\|x\|}
$$

Esto mide cuánto puede amplificar $A$ un vector $x$ en términos de la norma del vector.

#### Ejemplos:
- **Norma 1:**
  $$\|A\|_1 = \max_{1 \leq j \leq n} \sum_{i=1}^m |a_{ij}|$$
  Es el máximo de las sumas de los valores absolutos de las columnas de $A$.

- **Norma Infinita:**
  $$\|A\|_\infty = \max_{1 \leq i \leq m} \sum_{j=1}^n |a_{ij}|$$
  Es el máximo de las sumas de los valores absolutos de las filas de $A$.

- **Norma Euclidiana (o 2-norma):**
  $$\|A\|_2 = \sqrt{\lambda_{\text{máx}}(A^T A)}$$
  Donde $\lambda_{\text{máx}}$ es el valor propio máximo de $A^T A$.

---

### 2. **Norma de Frobenius**
Es una generalización de la norma euclidiana para matrices y se define como:

$$
\|A\|_F = \sqrt{\sum_{i=1}^m \sum_{j=1}^n |a_{ij}|^2}
$$

Esto equivale a tomar la raíz cuadrada de la suma de los cuadrados de todos los elementos de $A$. Se relaciona con el trazo de $A^T A$:

$$
\|A\|_F = \sqrt{\text{tr}(A^T A)}
$$

---


La norma de una matriz depende de la elección del tipo de norma. Cada una tiene propiedades particulares útiles para diferentes contextos en álgebra lineal, como estabilidad numérica, análisis de convergencia, o medidas de error.

In [ ]:
A = np.array([[1, -2, 3],
              [4, 5, -6],
              [-7, 8, 9]])

# Norma 1
norma_1 = np.max(np.sum(np.abs(A), axis=0))

# Norma infinita
norma_inf = np.max(np.sum(np.abs(A), axis=1))

# Norma de Frobenius
norma_frobenius = np.sqrt(np.sum(A**2))

# Norma 2 (máximo valor singular)
norma_2 = np.linalg.norm(A, 2)

print("Norma 1:", norma_1)
print("Norma infinita:", norma_inf)
print("Norma de Frobenius:", norma_frobenius)
print("Norma 2:", norma_2)


Norma 1: 18
Norma infinita: 24
Norma de Frobenius: 16.881943016134134
Norma 2: 14.428023632158306


# Matrices Complejas: Transpuesta y Conjugada Transpuesta

En álgebra lineal, una **matriz compleja** es una matriz cuyos elementos son números complejos, es decir, de la forma $a_{ij} = x + yi$, donde $x$ y $y$ son números reales e $i = \sqrt{-1}$.

A continuación, definimos los conceptos de transpuesta y conjugada transpuesta de una matriz compleja, con ejemplos prácticos.

## Definición de una Matriz Compleja
Consideremos la matriz $A$ de tamaño $2 \times 2$ como ejemplo:

$$
A = \begin{pmatrix}
1 + 2i & 3 - i \\
-2 + i & 4
\end{pmatrix}.
$$

## Transpuesta de una Matriz Compleja
La **transpuesta** de una matriz, denotada como $A^T$, se obtiene intercambiando las filas por las columnas. Es decir, el elemento en la posición $(i, j)$ de la matriz original pasa a la posición $(j, i)$ en la transpuesta.

Para la matriz $A$, su transpuesta es:

$$
A^T = \begin{pmatrix}
1 + 2i & -2 + i \\
3 - i & 4
\end{pmatrix}.
$$

### Pasos para Calcular la Transpuesta:
1. Identificar los elementos $a_{ij}$ de la matriz original.
2. Intercambiar las posiciones $(i, j)$ por $(j, i)$.
3. Reorganizar los elementos en la nueva disposición.

## Conjugada Transpuesta de una Matriz Compleja
La **conjugada transpuesta** de una matriz, también conocida como su **adjunta**, se denota como $A^*$ o $A^\dagger$. Se calcula aplicando dos operaciones consecutivas:
1. **Conjugado complejo:** Cambiar el signo de la parte imaginaria de cada elemento de la matriz.
2. **Transposición:** Intercambiar filas y columnas.

Para la matriz $A$, primero calculamos el conjugado complejo:

$$
\text{Conjugado complejo de } A =
\begin{pmatrix}
1 - 2i & 3 + i \\
-2 - i & 4
\end{pmatrix}.
$$

Luego, transponemos el resultado para obtener la conjugada transpuesta:

$$
A^* = \begin{pmatrix}
1 - 2i & -2 - i \\
3 + i & 4
\end{pmatrix}.
$$

---
Finalmente, dada la matriz $A$:

$$
A = \begin{pmatrix}
1 + 2i & 3 - i \\
-2 + i & 4
\end{pmatrix},
$$

- Su **transpuesta** es:

$$
A^T = \begin{pmatrix}
1 + 2i & -2 + i \\
3 - i & 4
\end{pmatrix}.
$$

- Su **conjugada transpuesta** es:

$$
A^\dagger = \begin{pmatrix}
1 - 2i & -2 - i \\
3 + i & 4
\end{pmatrix}.
$$


In [ ]:
A = np.array([
    [1+2j, 3 - 1j],
    [-2+1j, 4]
], dtype=complex)

transpuesta = A.T
conjugada = A.conjugate().T

print("Matriz Original:")
print(A)
print("\nTranspuesta:")
print(transpuesta)
print("\nConjugada Transpuesta:")
print(conjugada)

Matriz Original:
[[ 1.+2.j  3.-1.j]
 [-2.+1.j  4.+0.j]]

Transpuesta:
[[ 1.+2.j -2.+1.j]
 [ 3.-1.j  4.+0.j]]

Conjugada Transpuesta:
[[ 1.-2.j -2.-1.j]
 [ 3.+1.j  4.-0.j]]


# Matrices Simétricas y Hermitianas

Las matrices simétricas y hermitianas son tipos especiales de matrices cuadradas que presentan propiedades de simetría distintivas.

## Matriz Simétrica

### Propiedades
* **Definición**: Una matriz $A$ es simétrica si $A = A^T$
* Todos los elementos $a_{ij}$ son iguales a $a_{ji}$
* Válida solo para matrices de números reales

### Estructura
* Simétrica respecto a la diagonal principal
* Los elementos por encima y debajo de la diagonal son idénticos

$$
A = \begin{pmatrix}
a & b & c \\
b & d & e \\
c & e & f \\
\end{pmatrix}
$$

## Matriz Hermitiana

### Propiedades
* **Definición**: Una matriz $H$ es hermitiana si $H = H^\dagger$
* Válida para matrices de números complejos
* Los elementos $h_{ij}$ son el conjugado complejo de $h_{ji}$

### Características Específicas
* Elementos de la diagonal son números reales
* Elementos fuera de la diagonal son conjugados complejos entre sí

$$
A = \begin{pmatrix}
a & b + ci & d - ei \\
b - ci & f & g + hi \\
d + ei & g - hi & j \\
\end{pmatrix}
$$

## Diferencias Fundamentales

| Característica | Matriz Simétrica | Matriz Hermitiana |
|---------------|------------------|-------------------|
| Tipo de Números | Reales | Complejos |
| Condición de Simetría | $A = A^T$ | $H = H^\dagger$ |
| Elementos Diagonales | Cualquier número real | Solo números reales |



In [ ]:
A = np.array([
    [1, 2, 3],
    [2, 4, 5],
    [3, 5, 6]
])

trans = A.T

np.array_equal(A, trans)

True

In [ ]:
# Paso 1: Crear una matriz cuadrada aleatoria
n = 3  # Tamaño de la matriz
A = np.random.randint(0, 10, (n, n))

# Paso 2: Hacerla simétrica sumándola con su transpuesta y dividiendo por 2
A_simetrica = (A + A.T) / 2

A_simetrica_2 = A@A.T

# Mostrar la matriz original
print("Matriz original:")
print(A)
# Mostrar la matriz simétrica
print("\nMatriz simétrica:")
print(A_simetrica)

# Otra forma de matriz simétrica
print("\nMatriz simétrica 2:")
print(A_simetrica_2)

Matriz original:
[[8 9 3]
 [2 1 3]
 [3 3 6]]

Matriz simétrica:
[[8.  5.5 3. ]
 [5.5 1.  3. ]
 [3.  3.  6. ]]

Matriz simétrica 2:
[[154  34  69]
 [ 34  14  27]
 [ 69  27  54]]


In [ ]:
H = np.array([
    [1, 2 + 3j, 4 - 1j],
    [2 - 3j, 5, 6 + 2j],
    [4 + 1j, 6 - 2j, 7]
])

adj = H.conjugate().T

np.array_equal(H, adj)

True

In [ ]:
# Crear matriz compleja: parte real mas parte imaginaria
n = 3  # dimensión
H_re = np.random.randint(-10,10,size=(n,n)) # parte real
H_im = np.random.randint(-10,10,size=(n,n)) # parte imaginaria
H = H_re + 1j*H_im

sim = H@H.conj().T
print(sim)

[[202. +0.j 148.+12.j -81.+91.j]
 [148.-12.j 265. +0.j -99.+75.j]
 [-81.-91.j -99.-75.j  93. +0.j]]


### Importancia de Trabajar con Matrices Simétricas y Hermíticas

#### Matrices Simétricas
Las matrices simétricas son fundamentales en muchas áreas de las matemáticas y la ciencia debido a sus propiedades únicas y aplicaciones prácticas. Algunas de las razones más importantes para trabajar con matrices simétricas incluyen:

- **Propiedades Espectrales**: Las matrices simétricas tienen valores propios (eigenvalues) reales, lo cual es crucial en muchas aplicaciones físicas y de ingeniería.
- **Ortogonalidad**: Los vectores propios (eigenvectors) de matrices simétricas correspondientes a diferentes valores propios son ortogonales, lo que facilita los cálculos en algebra lineal y optimización.
- **Simplificación de Cálculos**: Las operaciones matemáticas y computacionales son más eficientes y estables cuando se trabaja con matrices simétricas debido a su estructura.
- **Aplicaciones en Física**: En mecánica cuántica y teoría de vibraciones, las matrices simétricas aparecen naturalmente en la descripción de sistemas físicos.

#### Matrices Hermíticas
Las matrices hermíticas son igualmente importantes, especialmente en campos que involucran números complejos y aplicaciones cuánticas. Algunas razones clave para trabajar con matrices hermíticas son:

- **Valores Propios Reales**: Al igual que las matrices simétricas, las matrices hermíticas tienen valores propios reales, lo que es vital para la estabilidad y realidad de las soluciones en problemas físicos.
- **Ortogonalidad Compleja**: Los vectores propios de matrices hermíticas correspondientes a diferentes valores propios son ortogonales con respecto al producto interior complejo, lo que facilita la análisis en espacios de Hilbert.
- **Teoría Cuántica**: En mecánica cuántica, los operadores hermíticos representan observables físicos, como la energía y el momento, asegurando que las mediciones sean reales.
- **Estabilidad Numérica**: Las matrices hermíticas poseen propiedades que garantizan la estabilidad numérica de algoritmos y métodos de solución en problemas de ingeniería y física.

Trabajar con matrices simétricas y hermíticas no solo simplifica los cálculos matemáticos, sino que también asegura que los resultados sean físicamente significativos y estables.


### Matriz Ortogonal
Una **matriz ortogonal** es una matriz cuadrada real $Q$ que satisface la relación $Q^T Q = Q Q^T = I$, donde $Q^T$ es la transpuesta de $Q$ e $I$ es la matriz identidad. En otras palabras, las columnas (y filas) de una matriz ortogonal son vectores ortonormales, lo que significa que son mutuamente ortogonales y tienen norma unitaria.

#### Propiedades de las Matrices Ortogonales:
- **Inversa**: La inversa de una matriz ortogonal es su transpuesta, es decir, $Q^{-1} = Q^T$.
- **Conservación de Longitudes**: Las transformaciones ortogonales conservan las longitudes de los vectores y los ángulos entre ellos.
- **Determinante**: El determinante de una matriz ortogonal es siempre $+1$ o $-1$.


In [ ]:
Q = np.array([
    [1/np.sqrt(2), 1/np.sqrt(2), 0],
    [-1/np.sqrt(2), 1/np.sqrt(2), 0],
    [0, 0, 1]
])
# Transpuesta
transpuesta = Q.T
# Inversa
inversa = np.linalg.inv(Q)
# Determinante
det = np.linalg.det(Q)

print("Matriz Original:")
print(Q)
print("\nTranspuesta:")
print(transpuesta)
print("\nInversa:")
print(inversa)
print("\nDeterminante:")
print(det)

Matriz Original:
[[ 0.70710678  0.70710678  0.        ]
 [-0.70710678  0.70710678  0.        ]
 [ 0.          0.          1.        ]]

Transpuesta:
[[ 0.70710678 -0.70710678  0.        ]
 [ 0.70710678  0.70710678  0.        ]
 [ 0.          0.          1.        ]]

Inversa:
[[ 0.70710678 -0.70710678  0.        ]
 [ 0.70710678  0.70710678  0.        ]
 [ 0.          0.          1.        ]]

Determinante:
0.9999999999999999


In [ ]:
# Comparación
print(np.isclose(transpuesta, inversa))
print(np.isclose(Q@Q.T, np.eye(len(Q))))

[[ True  True  True]
 [ True  True  True]
 [ True  True  True]]
[[ True  True  True]
 [ True  True  True]
 [ True  True  True]]


In [ ]:
# Usando linalg
sol = la.solve(Q, np.array([1,2,3]))

# Usando las propiedades
otra_sol = Q.T@np.array([1,2,3])

print(sol)
print(otra_sol)



[-0.70710678  2.12132034  3.        ]
[-0.70710678  2.12132034  3.        ]


### Matrices Unitarias
Las **matrices unitarias** son el análogo complejo de las matrices ortogonales. Una matriz cuadrada compleja $U$ es unitaria si satisface la relación $U^\dagger U = U U^\dagger = I$, donde $U^\dagger$ es la transpuesta conjugada de $U$. En otras palabras, una matriz unitaria conserva las normas y los ángulos en espacios de números complejos.

#### Propiedades de las Matrices Unitarias:
- **Inversa**: La inversa de una matriz unitaria es su transpuesta conjugada, es decir, $U^{-1} = U^\dagger$.
- **Conservación de Longitudes**: Las transformaciones unitarias conservan las normas de los vectores y los ángulos entre ellos.
- **Determinante**: El determinante de una matriz unitaria tiene módulo $1$.




In [ ]:
U = np.array([
    [1/np.sqrt(2), 1j/np.sqrt(2)],
    [1j/np.sqrt(2), 1/np.sqrt(2)]
])

transpuesta = U.conjugate().T
inversa = la.inv(U)
det = la.det(U)

print("Matriz Original:")
print(U)
print("\nTranspuesta:")
print(transpuesta)
print("\nInversa:")
print(inversa)
print("\nDeterminante:")
print(np.abs(det))



Matriz Original:
[[0.70710678+0.j         0.        +0.70710678j]
 [0.        +0.70710678j 0.70710678+0.j        ]]

Transpuesta:
[[0.70710678-0.j         0.        -0.70710678j]
 [0.        -0.70710678j 0.70710678-0.j        ]]

Inversa:
[[0.70710678+0.j         0.        -0.70710678j]
 [0.        -0.70710678j 0.70710678+0.j        ]]

Determinante:
0.9999999999999999


In [ ]:
np.isclose(transpuesta, inversa)

array([[ True,  True],
       [ True,  True]])

In [ ]:
sol = la.solve(U, np.array([1,2]))
sol_2 = U.conjugate().T@np.array([1,2])

print(sol)
print(sol_2)

[0.70710678-1.41421356j 1.41421356-0.70710678j]
[0.70710678-1.41421356j 1.41421356-0.70710678j]


### Relación entre Matrices Ortogonales y Unitarias:
- **Similitudes**: Ambas matrices conservan la longitud y los ángulos de los vectores, lo que las hace útiles en transformaciones geométricas y análisis de señales.
- **Diferencias**: Las matrices ortogonales operan en espacios reales, mientras que las matrices unitarias operan en espacios complejos.
- **Aplicaciones**: Las matrices ortogonales se utilizan en rotaciones y reflexiones en espacios euclidianos, mientras que las matrices unitarias son fundamentales en mecánica cuántica y procesamiento de señales.


# Valores Propios de una Matriz

Los valores propios (o eigenvalores) y los vectores propios (o eigenvectores) son conceptos fundamentales en álgebra lineal y tienen aplicaciones en diversas áreas como la física, la estadística y el análisis de sistemas.


Dada una matriz cuadrada $A$, un valor propio $\lambda$ es un número tal que existe un vector no nulo $v$ (llamado vector propio) que satisface la siguiente ecuación:

$$ A v = \lambda v $$

Esto significa que cuando la matriz $A$ actúa sobre el vector $v$, el resultado es el mismo vector $v$ multiplicado por el escalar $\lambda$.

## Cálculo de los Valores Propios

Para encontrar los valores propios de una matriz $A$, se resuelve la siguiente ecuación característica:

$$ \text{det}(A - \lambda I) = 0 $$

donde:
- $\text{det}$ denota el determinante de la matriz.
- $I$ es la matriz identidad del mismo tamaño que $A$.
- $\lambda$ es un escalar.

## Ejemplo

Consideremos la matriz:


$$ A = \begin{pmatrix} 4 & 1 \\ 2 & 3 \end{pmatrix} $$


Para encontrar los valores propios, resolvemos la ecuación característica:



$$ \text{det}(A - \lambda I) = \text{det}\begin{pmatrix} 4 - \lambda & 1 \\ 2 & 3 - \lambda \end{pmatrix} = 0 $$



Esto se expande como:



$$ (4 - \lambda)(3 - \lambda) - (1 \cdot 2) = \lambda^2 - 7\lambda + 10 = 0 $$



Resolviendo esta ecuación cuadrática, encontramos los valores propios:



$$ \lambda_1 = 5 $$




$$ \lambda_2 = 2 $$



Por lo tanto, los valores propios de la matriz $A$ son 5 y 2.


# Vector Propio

En álgebra lineal, un vector propio (o eigenvector) de una matriz $A$ es un vector no nulo que, al multiplicarlo por la matriz, se obtiene un escalar del mismo vector. Matemáticamente, un vector propio $v$ y un valor propio $\lambda$ cumplen la siguiente ecuación:



$$ A v = \lambda v $$



donde:
- $A$ es una matriz cuadrada.
- $v$ es un vector propio.
- $\lambda$ es un valor propio asociado a $v$.

## Ejemplo

Consideremos la matriz:



$$ A = \begin{pmatrix} 4 & 1 \\ 2 & 3 \end{pmatrix} $$



Para encontrar los valores propios de $A$, resolvemos la ecuación característica:



$$ \text{det}(A - \lambda I) = 0 $$





$$ \begin{vmatrix} 4 - \lambda & 1 \\ 2 & 3 - \lambda \end{vmatrix} = (4 - \lambda)(3 - \lambda) - 2 \cdot 1 = \lambda^2 - 7\lambda + 10 - 2 = \lambda^2 - 7\lambda + 8 = 0 $$



Resolviendo la ecuación cuadrática:



$$ \lambda^2 - 7\lambda + 8 = 0 $$





$$ \lambda = \frac{7 \pm \sqrt{49 - 32}}{2} = \frac{7 \pm 3}{2} $$





$$ \lambda_1 = 5, \quad \lambda_2 = 2 $$



Ahora, para encontrar los vectores propios asociados a estos valores propios:

1. Para $\lambda_1 = 5$:



$$ A v = 5 v $$





$$ \begin{pmatrix} 4 & 1 \\ 2 & 3 \end{pmatrix} \begin{pmatrix} x \\ y \end{pmatrix} = 5 \begin{pmatrix} x \\ y \end{pmatrix} $$





$$ \begin{cases} 4x + y = 5x \\ 2x + 3y = 5y \end{cases} $$



Resolviendo este sistema, obtenemos $ y = x $. Por lo tanto, un vector propio correspondiente a $\lambda_1 = 5$ es $ v_1 = \begin{pmatrix} 1 \\ 1 \end{pmatrix} $.

2. Para $\lambda_2 = 2$:



$$ A v = 2 v $$





$$ \begin{pmatrix} 4 & 1 \\ 2 & 3 \end{pmatrix} \begin{pmatrix} x \\ y \end{pmatrix} = 2 \begin{pmatrix} x \\ y \end{pmatrix} $$





$$ \begin{cases} 4x + y = 2x \\ 2x + 3y = 2y \end{cases} $$



Resolviendo este sistema, obtenemos $ y = -2x $. Por lo tanto, un vector propio correspondiente a $\lambda_2 = 2$ es $ v_2 = \begin{pmatrix} 1 \\ -2 \end{pmatrix} $.

## Resumen

Para la matriz $ A = \begin{pmatrix} 4 & 1 \\ 2 & 3 \end{pmatrix} $, los valores propios son $\lambda_1 = 5$ y $\lambda_2 = 2$, con los correspondientes vectores propios:

- Para $\lambda_1 = 5$: $ v_1 = \begin{pmatrix} 1 \\ 1 \end{pmatrix} $
- Para $\lambda_2 = 2$: $ v_2 = \begin{pmatrix} 1 \\ -2 \end{pmatrix} $


In [ ]:
A = np.array([
    [4, 1],
    [2, 3]
])

# Calcular los valores propios y vectores propios
valores_propios, vectores_propios = la.eig(A)
print("Valores propios:", valores_propios)
print("Vectores propios:\n", vectores_propios)

Valores propios: [5. 2.]
Vectores propios:
 [[ 0.70710678 -0.4472136 ]
 [ 0.70710678  0.89442719]]


# ¿Por qué los Vectores Propios son Diferentes?


La normalización es el proceso de escalar un vector de manera que tenga longitud 1. Si un vector propio manual se escala, mantiene la misma dirección pero cambia su magnitud. Por ejemplo:

- El vector propio manual $\begin{pmatrix} 1 \\ 1 \end{pmatrix}$ puede normalizarse a $\begin{pmatrix} \frac{1}{\sqrt{2}} \\ \frac{1}{\sqrt{2}} \end{pmatrix} \approx \begin{pmatrix} 0.70710678 \\ 0.70710678 \end{pmatrix}$.
- El vector propio manual $\begin{pmatrix} 1 \\ -2 \end{pmatrix}$ puede normalizarse a $\begin{pmatrix} \frac{1}{\sqrt{5}} \\ \frac{-2}{\sqrt{5}} \end{pmatrix} \approx \begin{pmatrix} 0.4472136 \\ -0.89442719 \end{pmatrix}$.





# Método de Potencias

El método de potencias es una técnica iterativa para encontrar el valor propio dominante (el de mayor magnitud) y su correspondiente vector propio de una matriz cuadrada $A$.

## Pasos del Método de Potencias

1. **Elección del Vector Inicial**:
   Escoge un vector inicial no nulo $x_0$. Este puede ser un vector aleatorio o una aproximación inicial del vector propio dominante.

2. **Iteración**:
   Itera utilizando la siguiente fórmula:
   
   

$$ x_{k+1} = \frac{A x_k}{\| A x_k \|} $$


   
   donde $\| A x_k \|$ es la norma (longitud) del vector $A x_k$.

3. **Convergencia**:
   Repite el paso anterior hasta que la secuencia de vectores $x_k$ converja. Esto ocurre cuando la dirección de $x_k$ deja de cambiar significativamente entre iteraciones.

4. **Cálculo del Valor Propio Dominante**:
   Una vez que el vector $x_k$ haya convergido a un vector propio $v$, el valor propio dominante $\lambda$ se puede aproximar utilizando:
   
   

$$ \lambda \approx \frac{(A v) \cdot v}{v \cdot v} $$


   
   donde $\cdot$ denota el producto escalar.

## Ejemplo

Supongamos que queremos encontrar el valor propio dominante de la matriz:



$$ A = \begin{pmatrix} 2 & 1 \\ 1 & 3 \end{pmatrix} $$



1. **Vector Inicial**:
   Escogemos un vector inicial, por ejemplo:
   
   

$$ x_0 = \begin{pmatrix} 1 \\ 1 \end{pmatrix} $$



2. **Iteraciones**:
   Realizamos las iteraciones necesarias. Por simplicidad, mostramos solo la primera iteración:

   

$$ x_1 = A x_0 = \begin{pmatrix} 2 & 1 \\ 1 & 3 \end{pmatrix} \begin{pmatrix} 1 \\ 1 \end{pmatrix} = \begin{pmatrix} 3 \\ 4 \end{pmatrix} $$


   
   Normalizamos $x_1$:

   

$$ x_1 = \frac{1}{\| Ax_0 \|} \begin{pmatrix} 3 \\ 4 \end{pmatrix} = \frac{1}{5} \begin{pmatrix} 3 \\ 4 \end{pmatrix} \approx \begin{pmatrix} 0.6 \\ 0.8 \end{pmatrix} $$



3. **Convergencia**:
   Continuamos iterando hasta que $x_k$ converge.

4. **Valor Propio Dominante**:
   Finalmente, calculamos el valor propio dominante $\lambda$ usando el vector propio $v$ obtenido.



In [ ]:
#@title Método de potencias
def metodo_potencias(A, max_iter=100, tolerancia=1e-10):
    """
    Implementación del método de potencias para encontrar el valor propio dominante
    y su vector propio asociado.

    Parámetros:
    A (numpy.ndarray): Matriz de entrada
    max_iter (int): Número máximo de iteraciones
    tolerancia (float): Criterio de convergencia

    Retorna:
    tuple: (valor_propio, vector_propio)
    """
    # Dimensión de la matriz
    n = A.shape[0]

    # Vector inicial de unos
    x_k = np.ones(n)

    for _ in range(max_iter):
        # Guardar el vector anterior para comparación
        x_k_anterior = x_k.copy()

        # Multiplicar la matriz por el vector
        x_k = A @ x_k

        # Normalizar el vector
        x_k = x_k / la.norm(x_k)

        # Criterio de convergencia
        if la.norm(x_k - x_k_anterior) < tolerancia:
            break

    # Calcular el valor propio dominante
    valor_propio = np.dot(x_k, A @ x_k) / np.dot(x_k, x_k)

    return valor_propio, x_k

In [ ]:
A = np.array([
    [4, 1],
    [2, 3]
])

valor_propio, vector_propio = metodo_potencias(A)
print(valor_propio)
print(vector_propio)

5.0
[0.70710678 0.70710678]


In [ ]:
A@vector_propio

array([3.53553391, 3.53553391])

In [ ]:
np.dot(valor_propio, vector_propio)

array([3.53553391, 3.53553391])

In [ ]:
np.isclose(la.det(A - valor_propio*np.eye(A.shape[0])),0)

np.True_

In [ ]:
np.isclose(la.norm(vector_propio), 1)

np.True_

## Método QR para Valores y Vectores Propios

El **método QR** es un procedimiento iterativo que permite calcular los valores y vectores propios de una matriz cuadrada $A$. Este método aprovecha las descomposiciones QR y la propiedad de que iterativamente $A_k$ converge hacia una matriz triangular superior, cuyos elementos diagonales son los valores propios de $A$.

---

### Descomposición QR
La descomposición QR de una matriz $A \in \mathbb{R}^{n \times n}$ es una factorización de la forma:

$$
A = Q R
$$

Donde:
- $Q$ es una matriz ortogonal (o unitaria si es compleja), es decir, $Q^T Q = I$.
- $R$ es una matriz triangular superior.

Esto puede calcularse mediante métodos como Gram-Schmidt, transformaciones de Householder, o rotaciones de Givens.

---

### Método Iterativo QR
El método QR para calcular los valores propios sigue estos pasos:

1. **Inicialización**:
   Comenzar con $A_0 = A$, la matriz original.

2. **Iteración**:
   Para $k = 0, 1, 2, \dots$:
   - Realizar la descomposición $A_k = Q_k R_k$, donde $Q_k$ es ortogonal y $R_k$ es triangular superior.
   - Actualizar $A_{k+1} = R_k Q_k$.

3. **Convergencia**:
   Repetir hasta que $A_k$ converja a una matriz triangular superior $T$, donde las entradas diagonales de $T$ son los valores propios de $A$.

---





## Ejemplo Numérico del Método QR

Consideremos la matriz:

$$
A = \begin{bmatrix}
4 & 1 \\
2 & 3
\end{bmatrix}
$$

Aplicaremos el método QR paso a paso para encontrar los valores propios.

---

### Paso 1: Descomposición QR de $A$
Para realizar la descomposición QR de $A$, necesitamos obtener $Q$ (matriz ortogonal) y $R$ (matriz triangular superior). Utilizaremos el proceso de **Gram-Schmidt**.

1. **Definimos las columnas de $A$ como vectores**:
   $$
   a_1 = \begin{bmatrix} 4 \\ 2 \end{bmatrix}, \quad
   a_2 = \begin{bmatrix} 1 \\ 3 \end{bmatrix}.
   $$

2. **Ortogonalizamos $a_2$ respecto a $a_1$**:
   - Normalizamos $a_1$ para obtener $q_1$:
    
$$ q_1 = \frac{a_1}{\|a_1\|} = \frac{\begin{bmatrix} 4 \\ 2 \end{bmatrix}}{\sqrt{4^2 + 2^2}} = \frac{\begin{bmatrix} 4 \\ 2 \end{bmatrix}}{\sqrt{20}} = \begin{bmatrix} \frac{2}{\sqrt{5}} \\ \frac{1}{\sqrt{5}} \end{bmatrix}.
$$

   - Proyección de $a_2$ sobre $q_1$:
$$
\text{proy}_{q_1}(a_2) = (q_1^T a_2) q_1 = \left(\begin{bmatrix} \frac{2}{\sqrt{5}} \\ \frac{1}{\sqrt{5}} \end{bmatrix}^T \begin{bmatrix} 1 \\ 3 \end{bmatrix}\right) \begin{bmatrix} \frac{2}{\sqrt{5}} \\ \frac{1}{\sqrt{5}} \end{bmatrix}.
$$

     Calculando:
$$
q_1^T a_2 = \frac{2}{\sqrt{5}} \cdot 1 + \frac{1}{\sqrt{5}} \cdot 3 = \frac{2 + 3}{\sqrt{5}} = \frac{5}{\sqrt{5}} = \sqrt{5}.
$$

     Entonces:
$$
\text{proy}_{q_1}(a_2) = \sqrt{5} \begin{bmatrix} \frac{2}{\sqrt{5}} \\ \frac{1}{\sqrt{5}} \end{bmatrix} = \begin{bmatrix} 2 \\ 1 \end{bmatrix}.
$$

   - Calculamos $q_2$ como:
$$
q_2 = \frac{a_2 - \text{proy}_{q_1}(a_2)}{\|a_2 - \text{proy}_{q_1}(a_2)\|} = \frac{\begin{bmatrix} 1 \\ 3 \end{bmatrix} - \begin{bmatrix} 2 \\ 1 \end{bmatrix}}{\sqrt{(-1)^2 + 2^2}} = \frac{\begin{bmatrix} -1 \\ 2 \end{bmatrix}}{\sqrt{5}} = \begin{bmatrix} -\frac{1}{\sqrt{5}} \\ \frac{2}{\sqrt{5}} \end{bmatrix}.
$$

3. **Construimos $Q$ y $R$**:
   - $Q = \begin{bmatrix} q_1 & q_2 \end{bmatrix} = \begin{bmatrix} \frac{2}{\sqrt{5}} & -\frac{1}{\sqrt{5}} \\ \frac{1}{\sqrt{5}} & \frac{2}{\sqrt{5}} \end{bmatrix}$.
   - Calculamos $R = Q^T A$:
$$
R = \begin{bmatrix} \frac{2}{\sqrt{5}} & \frac{1}{\sqrt{5}} \\ -\frac{1}{\sqrt{5}} & \frac{2}{\sqrt{5}} \end{bmatrix}^T \begin{bmatrix} 4 & 1 \\ 2 & 3 \end{bmatrix} = \begin{bmatrix} \sqrt{5} \cdot 4 & \sqrt{5} \cdot 3 \\ 0 & \sqrt{5} \cdot 2 \end{bmatrix}.
$$

---

### Paso 2: Primera Iteración
Actualizamos $A_1$ como:

$$
A_1 = RQ.
$$

Calculamos iterativamente hasta que $A_k$ converja a una matriz triangular superior. Los elementos diagonales de la matriz triangular son los **valores propios** de $A$. Por simplicidad, solo mostramos los resultados finales.

---

### Resultados
Después de varias iteraciones, obtenemos:

$$
A_k \approx \begin{bmatrix} 5 & 1 \\ 0 & 2 \end{bmatrix}.
$$

Los **valores propios** de $A$ son $5$ y $2$.

In [ ]:
#@title Método QR
def metodo_qr(A, tol=1e-10, max_iter=1000):
    n = A.shape[0]
    Ak = A.copy()
    for _ in range(max_iter):
        Q, R = la.qr(Ak)  # Descomposición QR
        Ak_next = R @ Q  # Actualización
        # Condición de parada: verificar convergencia diagonal
        if np.allclose(Ak, Ak_next, atol=tol):
            break
        Ak = Ak_next

    # Los valores propios están en la diagonal de Ak
    valores_propios = np.diag(Ak)
    return valores_propios, Ak




In [ ]:
A = np.array([[4, 1],
              [2, 3]])

valores, matriz_triangular = metodo_qr(A)
print("Valores propios:", valores)
print("Matriz triangular superior:", matriz_triangular)

Valores propios: [5. 2.]
Matriz triangular superior: [[5. 1.]
 [0. 2.]]
